js.go

```diff
@@ -3053,24 +3053,20 @@ func (sub *Subscription) Fetch(batch int, opts ...PullOpt) ([]*Msg, error) {
 		msgs = make([]*Msg, 0, batch)
 		msg  *Msg
 	)
-	for pmc && len(msgs) < batch {
+	// Clear all messages from prior fetch, since we're reusing the same message channel.
+	// If we reuse the 'cached' message during the next fetch call, we could get duplicated messages
+	// if the next fetch contains a redelivery for the same 'cached' messages.
+	for pmc {
 		// Check next msg with booleans that say that this is an internal call
 		// for a pull subscribe (so don't reject it) and don't wait if there
 		// are no messages.
-		msg, err = sub.nextMsgWithContext(ctx, true, false)
+		_, err = sub.nextMsgWithContext(ctx, true, false)
```

Comment:
I think this assumes that sub.Fetch is not being called concurrently right

Reply:
Hmmm.. suppose so. In which case this API is even more tricky. Because the same subscription is reused.

Reply:
I would try to add this first via an option first that ensures that it always discards previous results with a note that when used there should be always a single caller of Fetch within an app at any time with the side benefit that this prevents extra redeliveries.

Reply:
> I think this assumes that sub.Fetch is not being called concurrently right

Is `sub.Fetch` marked as thread safe? Does not seem the case.
When something is not marked thread-safe, it's probably not. Threads should not be sharing a subscription (unless they're locking it externally), so anyone doing it is probably experiencing worst issue than this one.

But maybe i missed it?

Reply:
`Subscription` has a lock, and `Fetch`, `nextMsgWithContext` use that lock as well. So can only assume it's thread-safe, even though it's not marked as such.

Reply:
That looks like an internal lock that could be guarding internal goroutines.
Doesn't mean the library is expected to be thread-safe to application threads.

(maybe it is, and it's just undocumented. But I don't think we can conclude 'this is thread safe externally' just because it has an internal lock)

Reply:
Poked around a bit. It does seem to be locking thoroughly to allow application threads safety. 
It's great that this is the case, would be nice to advertise it in documentation.

Apologies for the tangent.

Reply:
yes all the APIs are thread safe unless marked otherwise, concurrent Fetch and when it happens then the callers would be peeking from the shared internal buffer N messages.  I like the idea though of having a mode where each Fetch guarantees that there is nothing pending in the buffer every time it is called.

Reply:
Here is what the doc say:

>yes all the APIs are thread safe unless marked otherwise, 

https://go.dev/doc/comment#:~:text=By%20default%2C%20programmers%20should%20expect%20that%20a%20type%20is%20safe%20for%20use%20only%20by%20a%20single%20goroutine%20at%20a%20time.%20If%20a%20type%20provides%20stronger%20guarantees%2C%20the%20doc%20comment%20should%20state%20them.%20For%20example%3A


https://go.dev/doc/comment#:~:text=This%20example%20also,comments.%20For%20example%3A

(search for word "concurr" on this page)

Reply:
Thank you for finding this!

https://github.com/nats-io/nats.go/issues/1854
https://github.com/nats-io/nats.go/issues/1855